src/dbnode/environment/config.go

```diff
@@ -83,19 +85,84 @@ type SeedNodeSecurityConfig struct {
 	AutoTLS       bool   `yaml:"autoTls"`
 }
 
+// DynamicConfiguration is used for running M3DB with a dynamic config
+type DynamicConfiguration []*DynamicCluster
+
+// DynamicCluster is a single cluster in a dynamic configuration
+type DynamicCluster struct {
+	Async   bool                      `yaml:"async"`
+	Service *etcdclient.Configuration `yaml:"service"`
+}
+
+// Validate validates the DynamicConfiguration.
+func (c DynamicConfiguration) Validate() error {
+	syncCount := 0
+	for _, cfg := range c {
+		if !cfg.Async {
+			syncCount++
+		}
+	}
+	if syncCount != 1 {
+		return errInvalidSyncCount
+	}
+	return nil
+}
+
+// SyncCluster returns the synchronous cluster in the DynamicConfiguration
+func (c DynamicConfiguration) SyncCluster() (*DynamicCluster, error) {
+	for _, cluster := range c {
+		if !cluster.Async {
```

Comment:
Maybe before checking, should `c.Validate()` and return error if it returns error?  Otherwise it's possible probably for code paths to call `SyncCluster()` and just get the first non-async one (which is invalid).